test/cluster/test_automatic_cleanup.py

```diff
@@ -68,3 +72,99 @@ async def test_no_cleanup_when_unnecessary(manager: ManagerClient):
+    """Scenario:
+       * Start two nodes, a vnodes-based table with an rf=2
+       * Run insert while bootstrapping another node, suspend this insert in database_apply_wait injection
+       * Bootstrap succeeds, capture the final version
+       * Start decommission -> triggers global barrier, which we fail on another injection
+       * This failure is not fatal, the cleanup procedure continues and blocks on waiting for the stale write
+       * We release the database_apply_wait injection, cleanup succeeds, write fails with 'stale topology exception'
+    """
+
+    config = {'tablets_mode_for_new_keyspaces': 'disabled'}
+
+    logger.info("start first server")
+    servers = [await manager.server_add(property_file={"dc": "dc1", "rack": "rack1"}, config=config)]
+    servers += [await manager.server_add(property_file={"dc": "dc1", "rack": "rack2"}, config=config)]
+
+    (cql, hosts) = await manager.get_ready_cql(servers)
+    log0 = await manager.server_open_log(servers[0].server_id)
+    log1 = await manager.server_open_log(servers[1].server_id)
+
+    async with new_test_keyspace(manager, "WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 2}") as ks:
+        logger.info("Create table my_test_table")
+        await cql.run_async(f"CREATE TABLE {ks}.my_test_table (pk int PRIMARY KEY, c int);")
+
+        # Have a bootstrapping node hang in write_both_read_new
+        logger.info("Add third node")
+        servers += [await manager.server_add(
+            property_file={"dc": "dc1", "rack": "rack3"},
+            config=config,
+            start=False
+        )]
+        logger.info("Enable 'topology_coordinator/write_both_read_new/after_barrier' injection")
+        await manager.api.enable_injection(servers[0].ip_addr,
+                                           'topology_coordinator/write_both_read_new/after_barrier',
+                                           True)
+        logger.info("Start bootstrapping the third node")
+        bootstrap_task = asyncio.create_task(manager.server_start(servers[2].server_id))
+        logger.info("Waiting for topology_coordinator/write_both_read_new/after_barrier")
+        await log0.wait_for("topology_coordinator/write_both_read_new/after_barrier: waiting for message")
+
+        # Have a write request with write_both_read_new version stuck on both nodes:
+        # - On the first node, this exercises the coordinator fencing code path.
+        # - On the second node, this exercises the replica code path.
+        logger.info("Enable 'database_apply_wait' injection")
+        for s in servers[:-1]:
+            await manager.api.enable_injection(s.ip_addr, 'database_apply_wait',
+                                               False, parameters={'cf_name': 'my_test_table'})
+        logger.info("Start write")
+        write_task = cql.run_async(f"INSERT INTO {ks}.my_test_table (pk, c) VALUES (1, 1)", host=hosts[0])
```

Comment:
Do we make sure that the write is fenced out on both replicas? You only check that the write fails with "stale topology exception". If the write uses CL=2, that's not enough. And I guess that the default CL in the driver is QUORUM so 2 in this case. I think you should write with CL=1.

Alternatively, you can look for "stale topology exception" in the logs of both replicas. Or even do both -- write with CL=1 and look for logs.